### PR TITLE
close #65 - Remove location model

### LINKED ISSUE #64: Can't create new devices
##### Detail:
When I send request to create a device, I receive Internal Server Error.

##### Curl command to create a device:
```bash
curl -X POST http://api-proxy/devices \
     -H 'Authorization: Bearer d67195569fb3fa2acedf8e318c442fd1b9fb47f8d109181be215b01e62029d5e' \
     -H 'Content-Type: application/json' \
     -d '{
           "name": "Alarma 1",
           "type": { "id": "58890ffce8cc35347a000001" }
         }'
```

##### Proxy log:
```bash
nginx-proxy_1    | nginx.1    | api-proxy 172.19.0.1 - - [25/Jan/2017:21:10:43 +0000] "POST /devices HTTP/1.1" 500 19429 "-" "curl/7.47.0"
devices_1        | 21:10:43 web.1  | 172.19.0.1,172.19.0.3 - - [25/Jan/2017 21:10:43] "POST /devices HTTP/1.1" 500 19429 0.0764
```

##### Device microservice logs:
```bash
devices_1        | 21:10:43 web.1  |   MOPED: 172.19.0.5:27017 QUERY        database=people collection=oauth_access_tokens selector={"$query"=>{"token"=>"d67195569fb3fa2acedf8e318c442fd1b9fb47f8d109181be215b01e62029d5e"}, "$orderby"=>{:_id=>1}} flags=[:slave_ok] limit=-1 skip=0 batch_size=nil fields=nil (1.4741ms)
devices_1        | 21:10:43 web.1  |   MOPED: 172.19.0.5:27017 QUERY        database=people collection=users selector={"_id"=>"58890d8685b9b33aa7000001"} flags=[:slave_ok] limit=0 skip=0 batch_size=nil fields=nil (1.2710ms)
devices_1        | 21:10:43 web.1  |
devices_1        | 21:10:43 web.1  |
devices_1        | 21:10:43 web.1  | Started POST "/devices" for 172.19.0.15 at 2017-01-25 21:10:43 +0000
devices_1        | 21:10:43 web.1  | Processing by DevicesController#create as JSON
devices_1        | 21:10:43 web.1  |   Parameters: {"name"=>"Alarma 1", "type"=>{"id"=>"58890ffce8cc35347a000001"}}
devices_1        | 21:10:43 web.1  |   MOPED: 172.19.0.5:27017 QUERY        database=people collection=oauth_access_tokens selector={"$query"=>{"token"=>"d67195569fb3fa2acedf8e318c442fd1b9fb47f8d109181be215b01e62029d5e"}, "$orderby"=>{:_id=>1}} flags=[:slave_ok] limit=-1 skip=0 batch_size=nil fields=nil (0.9303ms)
devices_1        | 21:10:43 web.1  |   MOPED: 172.19.0.5:27017 QUERY        database=people collection=users selector={"_id"=>"58890d8685b9b33aa7000001"} flags=[:slave_ok] limit=0 skip=0 batch_size=nil fields=nil (0.4950ms)
devices_1        | 21:10:43 web.1  | WARNING: Can't mass-assign protected attributes: format, action, controller
devices_1        | 21:10:43 web.1  | Completed 500 Internal Server Error in 13ms
devices_1        | 21:10:43 web.1  |
devices_1        | 21:10:43 web.1  | NoMethodError (undefined method `match' for nil:NilClass):
devices_1        | 21:10:43 web.1  |   app/models/device.rb:86:in `touch_locations'
devices_1        | 21:10:43 web.1  |   app/controllers/devices_controller.rb:32:in `create'
devices_1        | 21:10:43 web.1  |   lib/middlewares/hostable.rb:8:in `call'
devices_1        | 21:10:43 web.1  |   lib/middlewares/daily_rate_limit.rb:8:in `call'
devices_1        | 21:10:43 web.1  |
devices_1        | 21:10:43 web.1  |
devices_1        | 21:10:43 web.1  |   Rendered /var/lib/gems/1.9.1/gems/actionpack-3.2.13/lib/action_dispatch/middleware/templates/rescues/_trace.erb (1.1ms)
devices_1        | 21:10:43 web.1  |   Rendered /var/lib/gems/1.9.1/gems/actionpack-3.2.13/lib/action_dispatch/middleware/templates/rescues/_request_and_response.erb (0.8ms)
devices_1        | 21:10:43 web.1  |   Rendered /var/lib/gems/1.9.1/gems/actionpack-3.2.13/lib/action_dispatch/middleware/templates/rescues/diagnostics.erb within rescues/layout (18.2ms)
```

##### Type _**58890ffce8cc35347a000001**_ exists:
```bash
curl http://api-proxy/types  \
    -H 'Authorization: Bearer d67195569fb3fa2acedf8e318c442fd1b9fb47f8d109181be215b01e62029d5e'

[{"id":"58891181e8cc35347a000002","uri":"http://api-proxy/types/58891181e8cc35347a000002","name":"Alarma 1","category":null,"owner":{"id":"58890d8685b9b33aa7000001","uri":"http://api-proxy/people/58890d8685b9b33aa7000001"},"description":"","created_at":"2017-01-25T20:58:41Z","updated_at":"2017-01-25T20:58:41Z"},{"id":"58890ffce8cc35347a000001","uri":"http://api-proxy/types/58890ffce8cc35347a000001","name":"Light","category":null,"owner":{"id":"58890d8685b9b33aa7000001","uri":"http://api-proxy/people/58890d8685b9b33aa7000001"},"description":"","created_at":"2017-01-25T20:52:12Z","updated_at":"2017-01-25T20:52:12Z"}]
```

This error are occurred in this line: [link](https://github.com/lelylan/devices/blob/master/app/models/device.rb#L86)

### LINKED ISSUE #65: Remove location model
-> lelylan/devices#64

### LINKED ISSUE #67: RuntimeError in DevicesController#index
#### Information
When i go to: [http://api-proxy.lelylan.dev.exoip.com.ar/devices?access_token=f7139bbdd917168f71d5b13cc6ae3e215a0e4a86f45ba137c16dca3c8a588572&token_type=bearer&expires_in=86400&state=remember-me](http://api-proxy.lelylan.dev.exoip.com.ar/devices?access_token=f7139bbdd917168f71d5b13cc6ae3e215a0e4a86f45ba137c16dca3c8a588572&token_type=bearer&expires_in=86400&state=remember-me)
The api respond: **Called id for nil, which would mistakenly be 4 -- if you really wanted the id of nil, use object_id**.

#### Response page:
captura de pantalla de 2017-02-06 17-19-50

#### Error detail:
This error is generated from : `app/controllers/devices_controller.rb:68:in 'find_owned_resources'`

##### In **app/controllers/devices_controller.rb**:
```ruby
66: def find_owned_resources
67:   @devices = Device.where(resource_owner_id: current_user.id)
68: end
```

##### In **app/controllers/application_controller.rb**:
```ruby
7: def current_user
8:   @current_user ||= User.find(@device.resource_owner_id)          if request.headers['X-Physical-Secret']
9:   @current_user ||= User.find(doorkeeper_token.resource_owner_id) if request.headers['Authorization']
10:   return @current_user
11: end
```

###### Variables used in **current_user**:
```
@device.resource_owner_id -> nil
request.headers['X-Physical-Secret'] -> nil
doorkeeper_token.resource_owner_id -> 5894d5c59f6fb8bb72000001d
request.headers['Authorization'] -> nil
```

The variable **doorkeeper_token.resource_owner_id** has correct user id **5894d5c59f6fb8bb72000001d** but **request.headers['Authorization']** is nil, then, **@current_user** not set in **current_user** function

### LINKED ISSUE #68: RuntimeError in DevicesController#show
#### Information
When i go to: [`http://api-proxy.lelylan.dev.exoip.com.ar/devices/5898c7fc6899264eba000001?access_token=084539f2401f6e7ced4a0ffc822c5ba7e4cb3ea4ba5f249bbb7d0be99a6e23f0&token_type=bearer&expires_in=86400&state=remember-me`](`http://api-proxy.lelylan.dev.exoip.com.ar/devices/5898c7fc6899264eba000001?access_token=084539f2401f6e7ced4a0ffc822c5ba7e4cb3ea4ba5f249bbb7d0be99a6e23f0&token_type=bearer&expires_in=86400&state=remember-me`)
The api respond: **Called id for nil, which would mistakenly be 4 -- if you really wanted the id of nil, use object_id**.

#### Response page:
captura de pantalla de 2017-02-06 16-58-57

#### Error detail:
This error is generated from : `app/controllers/devices_controller.rb:68:in 'find_owned_resources'`

##### In **app/controllers/devices_controller.rb**:
```ruby
66: def find_owned_resources
67:   @devices = Device.where(resource_owner_id: current_user.id)
68: end
```

##### In **app/controllers/application_controller.rb**:
```ruby
7: def current_user
8:   @current_user ||= User.find(@device.resource_owner_id)          if request.headers['X-Physical-Secret']
9:   @current_user ||= User.find(doorkeeper_token.resource_owner_id) if request.headers['Authorization']
10:   return @current_user
11: end
```

###### Variables used in **current_user**:
```
@device.resource_owner_id -> nil
request.headers['X-Physical-Secret'] -> nil
doorkeeper_token.resource_owner_id -> 5894d5c59f6fb8bb72000001d
request.headers['Authorization'] -> nil
```

The variable **doorkeeper_token.resource_owner_id** has correct user id **5894d5c59f6fb8bb72000001d** but **request.headers['Authorization']** is nil, then, **@current_user** not set in **current_user** function

### PR DESCRIPTION
**Pull request by FedeG**
Close #64 
Fix #67
Fix #68

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class ApplicationController < ActionController::API
 
   def current_user
     @current_user ||= User.find(@device.resource_owner_id)          if request.headers['X-Physical-Secret']
-    @current_user ||= User.find(doorkeeper_token.resource_owner_id) if request.headers['Authorization']
+    @current_user ||= User.find(doorkeeper_token.resource_owner_id) if request.headers['Authorization'] or (not request.headers['X-Physical-Secret'] and doorkeeper_token)
```
Comment (andreareginato) [inline app/controllers/application_controller.rb]: I would try to clean it up a bit more. 
This way the `'X-Physical-Secret` token whenever it is set, will have priority

```
...
@current_user ||=  User.find(doorkeeper_token.resource_owner_id) if doorkeeper_token or request.headers['Authorization']
@current_user ||= User.find(@device.resource_owner_id) if request.headers['X-Physical-Secret']
```

**Review by andreareginato — CHANGES_REQUESTED** [review]
When the refactoring in `application_controller` is done we can merge this.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (FedeG) [inline app/controllers/application_controller.rb]: -> https://github.com/lelylan/devices/pull/66/commits/e2d1db8ba04c531a055e9d5e2a8b80f584f43581

**Comment by andreareginato** [comment]
Perfect work.

**Pull request merged**